lib/pure/asyncnet.nim

```diff
@@ -822,40 +822,44 @@ proc sendTo*(socket: AsyncSocket, address: string, port: Port, data: string,
     else:
       raise newException(IOError, "Couldn't resolve address: " & address)
 
-proc recvFrom*(socket: AsyncSocket, size: int,
-               flags = {SocketFlag.SafeDisconn}):
-              owned(Future[tuple[data: string, address: string, port: Port]])
+proc recvFrom*(socket: AsyncSocket, data: FutureVar[string], size: int,
+               address: FutureVar[string], port: FutureVar[Port],
+               flags = {SocketFlag.SafeDisconn}): owned(Future[int])
               {.async, since: (1, 3).} =
-  ## Receives a datagram data from ``socket``, which must be at least of size
-  ## ``size``. Returned future will complete once one datagram has been received
-  ## and will return tuple with: data of packet received; and address and port
-  ## of datagram's sender.
-  ## 
+  ## Receives a datagram data from ``socket`` into ``data``, which must be at
+  ## least of size ``size``. The address and port of datagram's sender will be
+  ## stored into ``address`` and ``port``, respectively. Returned future will
+  ## complete once one datagram has been received, and will return size of
+  ## packet received.
+  ##
   ## If an error occurs an OSError exception will be raised.
   ## 
   ## This proc is normally used with connectionless sockets (UDP sockets).
+  ## 
+  ## **Note**: ``data`` must be initialized to the length of ``size``.
   template adaptRecvFromToDomain(domain: Domain) =
     var lAddr = sizeof(sAddr).SockLen
     
-    let fut = await recvFromInto(AsyncFD(getFd(socket)), cstring(data), size,
-                                 cast[ptr SockAddr](addr sAddr), addr lAddr,
-                                 flags)
+    result = await recvFromInto(AsyncFD(getFd(socket)), cstring(data.mget()), size,
+                                cast[ptr SockAddr](addr sAddr), addr lAddr,
+                                flags)
     
-    data.setLen(fut)
-    
-    result.data = data
-    result.address = getAddrString(cast[ptr SockAddr](addr sAddr))
+    data.mget().setLen(result)
+    data.complete()
+
+    address.mget.add(getAddrString(cast[ptr SockAddr](addr sAddr)))
```

Comment:
This isn't right. You'll be adding new data into the ``address`` every time this is called. Instead you should ask the caller to prealloc the maximum IPv6 length and do the same as you do for `data`.

Reply:
The problem with pre-allocating `address` is that `getAddrString()` will allocate another string and not write the IP address in the memory already allocated in `address`. And to pass the result of `getAddrString()` the memory allocated in `address` will be discarded. That is, this pre-allocation is unnecessary.

`address.mget.add()` is in fact not the best solution, as it will add new data in `address` to each call, when the user does not send an empty variable. However, you can add an assertion to check that `0 == len(address.mget())`.

Another way would be the one that was used first and you didn't like it: `address.mget() = getAddrString(cast[ptr SockAddr](addr sAddr))`

In my view, the only way to pre-allocate `address` and be able to write to that memory would be to redo `getAddrString()`.

@dom96, what's wrong with using: `address.mget() = getAddrString(cast[ptr SockAddr](addr sAddr))`? Could you explain it to me, because I'm a bit of a layman on the subject?

Reply:
The whole point of this overload is to reduce string allocations. Calling ``getAddrString`` will allocate a string, so you do need to create an overload for it to take a `var string` as well and use it here IMO.

Reply:
I hope that's what you expect.